aten/src/ATen/templates/TensorBody.h

```diff
@@ -976,6 +976,57 @@ struct MaybeOwnedTraits<at::Tensor> {
     return true;
   }
 };
+
+template <>
+struct ExclusivelyOwnedTraits<at::Tensor> {
+  using repr_type = at::Tensor;
+  using pointer_type = const at::Tensor*;
+
+  static repr_type nullRepr() {
+    // REVIEW: do we need to be able to distinguish an empty
+    // ExclusivelyOwned<Tensor> from an undefined Tensor?
+    return at::Tensor();
+  }
+
+  template <class... Args>
+  static repr_type createInPlace(Args&&... args) {
+    return at::Tensor(std::forward<Args>(args)...);
+  }
+
+  static repr_type moveToRepr(at::Tensor&& x) {
+    return std::move(x);
+  }
+
+  static void destroyOwned(at::Tensor& x) {
+    TensorImpl*const toDestroy = x.unsafeReleaseTensorImpl();
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(toDestroy != nullptr, "Tensor somehow got null TensorImpl?");
+    // May be 0 because UndefinedTensorImpl doesn't get its refcount
+    // incremented.
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
+        toDestroy->refcount_ == 1 || (toDestroy->refcount_ == 0 && toDestroy == UndefinedTensorImpl::singleton()),
+        "ExclusivelyOwned<Tensor> destroyed with refcount ", toDestroy->refcount_, ", expected 1!");
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
+        toDestroy->weakcount_ == 1 || (toDestroy->weakcount_ == 0 && toDestroy == UndefinedTensorImpl::singleton()),
+        "ExclusivelyOwned<Tensor> destroyed with weakcount ", toDestroy->weakcount_, ", expected 1!");
+    if (toDestroy != UndefinedTensorImpl::singleton()) {
+#ifndef NDEBUG
+      // Needed to pass the debug assertions in ~intrusive_ptr_target.
+      toDestroy->refcount_ = 0;
+      toDestroy->weakcount_ = 0;
```

Comment:
Hmm, this is extra invariant on `release_resources` that the refcount/weakcount may not be accurate

Reply:
if release_resources cares about the refcount/weakcount....why?
We can do the stores in all builds if you think it will make a difference to anyone, I would just rather not have the code size and memory traffic.

Reply:
oh, wait, it *can't* care about the refcount & weakcount because they're private to intrusive_ptr_target!

Reply:
I don't think `release_resources` ever relies on it, but it's a near miss, as some PyObject binding stuff uses `use_count` to decide if it has exclusive ownership or not. Just need to doc and socialize this, no action needed.